File: src/archivist.ts
/**
 * Archivist — append-only audit store outside the Circulator cycle.
 *
 * Circulator memory M is a *cycling* store: active → pruned → memory →
 * retrieved → active. Archivist records content that should leave that
 * loop permanently: a write-once log of what was pruned, when, and under
 * what score — for audit / compliance, not automatic reuse every cycle.
 *
 * Guarantees (sketch):
 *   - append-only: no update-in-place, no silent delete of records
 *   - deliberate retrieval only (by id, hash, time range, reason)
 *   - distinct from Circulator's working memory
 *
 * @see Future Work in README — Archivist and Originist
 */

export type ArchiveReason = "pruned" | "overflow" | "budget" | "manual" | "demoted";

export type ArchiveClassification = "fact" | "event" | "instruction" | "task";

export interface ArchiveRecord {
  /** Stable record id (unique per append). */
  id: string;
  content_hash: string;
  /** Content snapshot at archive time (may be residual / truncated). */
  residual: string;
  /** Score at the moment of exit from the active loop. */
  score: number;
  classification: ArchiveClassification;
  reason: ArchiveReason;
  session_id: string;
  agent_type: string;
  timestamp_ms: number;
  /** Originist generation count at archive time, if known. */
  generation?: number;
  /** Optional free-form metadata (threshold, agent role, etc.). */
  meta?: Record<string, string | number | boolean>;
}

export interface ArchiveInput {
  content: string;
  content_hash?: string;
  score?: number;
  classification?: ArchiveClassification;
  reason?: ArchiveReason;
  session_id?: string;
  agent_type?: string;
  timestamp_ms?: number;
  generation?: number;
  meta?: Record<string, string | number | boolean>;
}

export interface ArchivistQuery {
  id?: string;
  content_hash?: string;
  reason?: ArchiveReason;
  session_id?: string;
  /** Inclusive lower bound (ms). */
  since_ms?: number;
  /** Inclusive upper bound (ms). */
  until_ms?: number;
  /** Max records to return (newest first). Default unlimited. */
  limit?: number;
}

export interface ArchivistOptions {
  /** Soft cap for in-memory records. When exceeded, oldest stay on disk only if path set. */
  cap?: number;
  /** Optional append-only JSONL path for durable audit log. */
  path?: string;
}

function simpleHash(s: string): string {
  let h = 0;
  for (let i = 0; i < s.length; i++) {
    h = ((h << 5) - h + s.charCodeAt(i)) | 0;
  }
  return `h-${Math.abs(h).toString(36)}`;
}

function newRecordId(hash: string, ts: number, seq: number): string {
  return `arch-${hash}-${ts}-${seq}`;
}

/**
 * Append-only audit store. Records never mutate after append.
 * `clear()` is test-only and does not erase durable JSONL history.
 */
export class Archivist {
  private records: ArchiveRecord[] = [];
  private seq = 0;
  private readonly cap: number;
  private readonly path?: string;

  constructor(options: ArchivistOptions = {}) {
    this.cap = options.cap ?? 10_000;
    this.path = options.path;
  }

  /** Append a permanent exit record. Returns the stored record. */
  record(input: ArchiveInput): ArchiveRecord {
    const timestamp_ms = input.timestamp_ms ?? Date.now();
    const content_hash = input.content_hash ?? simpleHash(input.content);
    this.seq += 1;
    const entry: ArchiveRecord = {
      id: newRecordId(content_hash, timestamp_ms, this.seq),
      content_hash,
      residual: input.content,
      score: input.score ?? 0,
      classification: input.classification ?? "fact",
      reason: input.reason ?? "pruned",
      session_id: input.session_id ?? "unknown",
      agent_type: input.agent_type ?? "kompress",
      timestamp_ms,
      generation: input.generation,
      meta: input.meta,
    };
    // Freeze conceptually: store a shallow copy; never reassign fields later.
    this.records.push({ ...entry, meta: entry.meta ? { ...entry.meta } : undefined });
    this.appendDurable(entry);
    if (this.records.length > this.cap) {
      // Keep memory bounded; durable log (if any) retains full history.
      this.records.splice(0, this.records.length - this.cap);
    }
    return entry;
  }

  /** Batch append. */
  recordMany(inputs: ArchiveInput[]): ArchiveRecord[] {
    return inputs.map((i) => this.record(i));
  }

  size(): number {
    return this.records.length;
  }

  /**
   * Deliberate retrieval — not called on every compress cycle.
   * Filters are AND-ed; results newest-first.
   */
  query(q: ArchivistQuery = {}): ArchiveRecord[] {
    let out = this.records.slice();
    if (q.id !== undefined) out = out.filter((r) => r.id === q.id);
    if (q.content_hash !== undefined) out = out.filter((r) => r.content_hash === q.content_hash);
    if (q.reason !== undefined) out = out.filter((r) => r.reason === q.reason);
    if (q.session_id !== undefined) out = out.filter((r) => r.session_id === q.session_id);
    if (q.since_ms !== undefined) out = out.filter((r) => r.timestamp_ms >= q.since_ms!);
    if (q.until_ms !== undefined) out = out.filter((r) => r.timestamp_ms <= q.until_ms!);
    out.sort((a, b) => b.timestamp_ms - a.timestamp_ms);
    if (q.limit !== undefined && q.limit >= 0) out = out.slice(0, q.limit);
    return out.map((r) => ({ ...r, meta: r.meta ? { ...r.meta } : undefined }));
  }

  /** Full in-memory snapshot (audit export). Does not include spilled-only history. */
  snapshot(): ArchiveRecord[] {
    return this.query();
  }

  /** Test helper — clears in-memory only; durable JSONL is not truncated. */
  clearMemory(): void {
    this.records = [];
  }

  private appendDurable(entry: ArchiveRecord): void {
    if (!this.path) return;
    try {
      const line = JSON.stringify(entry) + "\n";
      // Best-effort append via Bun when available; silent on failure.
      const bun = (globalThis as { Bun?: { write: (p: string, d: string | Uint8Array) => Promise<number>; file: (p: string) => { size: number; arrayBuffer: () => Promise<ArrayBuffer> } } }).Bun;
      if (!bun) return;
      void (async () => {
        try {
          let existing = new Uint8Array(0);
          try {
            const f = bun.file(this.path!);
            if (f.size > 0) existing = new Uint8Array(await f.arrayBuffer());
          } catch {
            // new file
          }
          const enc = new TextEncoder().encode(line);
          const merged = new Uint8Array(existing.length + enc.length);
          merged.set(existing);
          merged.set(enc, existing.length);
          await bun.write(this.path!, merged);
        } catch {
          // silent — audit file best-effort
        }
      })();
    } catch {
      // silent
    }
  }
}

export function createArchivist(options?: ArchivistOptions): Archivist {
  return new Archivist(options);
}

/**
 * Helper: archive dropped messages from a compress pass (exit from active C).
 * Does not enqueue Circulator — that remains a separate, cycling path.
 */
export function archiveDropped(
  archivist: Archivist,
  dropped: Array<{ content: string; score?: number; role?: string }>,
  opts: {
    reason?: ArchiveReason;
    session_id?: string;
    agent_type?: string;
    threshold?: number;
  } = {},
): ArchiveRecord[] {
  return dropped.map((d) =>
    archivist.record({
      content: d.content,
      score: d.score ?? 0,
      reason: opts.reason ?? "pruned",
      session_id: opts.session_id,
      agent_type: opts.agent_type,
      meta: {
        ...(opts.threshold !== undefined ? { threshold: opts.threshold } : {}),
        ...(d.role !== undefined ? { role: d.role } : {}),
      },
    }),
  );
}

File: crates/kompress-core/src/archivist.rs
//! Archivist — append-only audit store outside the Circulator cycle.
//!
//! Circulator's memory is a *cycling* store (active → pruned → memory →
//! retrieved → active). Archivist records permanent exits from that loop:
//! what was pruned, when, and under what score — for audit, not automatic
//! per-cycle retrieval.
//!
//! Sketch status: in-memory append-only log with deliberate query API.
//! Durable backends (JSONL, object store) can wrap `record` without changing
//! the role contract.

use crate::types::ContextUnit;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ArchiveReason {
    Pruned,
    Overflow,
    Budget,
    Manual,
    Demoted,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ArchiveRecord {
    pub id: String,
    pub content_hash: String,
    pub residual: String,
    pub score: f64,
    pub reason: ArchiveReason,
    pub timestamp_ms: i64,
    pub generation: Option<u32>,
    pub is_critical_syntactic: bool,
    pub token_count: usize,
}

#[derive(Debug, Clone, Default)]
pub struct ArchivistQuery {
    pub id: Option<String>,
    pub content_hash: Option<String>,
    pub reason: Option<ArchiveReason>,
    pub since_ms: Option<i64>,
    pub until_ms: Option<i64>,
    pub limit: Option<usize>,
}

#[derive(Debug, Clone)]
pub struct Archivist {
    records: Vec<ArchiveRecord>,
    seq: u64,
    cap: usize,
}

impl Archivist {
    pub fn new() -> Self {
        Self {
            records: Vec::new(),
            seq: 0,
            cap: 10_000,
        }
    }

    pub fn with_cap(cap: usize) -> Self {
        Self {
            records: Vec::new(),
            seq: 0,
            cap: cap.max(1),
        }
    }

    /// Append a permanent exit record. Never updates existing rows.
    pub fn record(
        &mut self,
        residual: impl Into<String>,
        score: f64,
        reason: ArchiveReason,
        timestamp_ms: i64,
        generation: Option<u32>,
        is_critical_syntactic: bool,
        token_count: usize,
    ) -> ArchiveRecord {
        let residual = residual.into();
        let content_hash = simple_hash(&residual);
        self.seq += 1;
        let entry = ArchiveRecord {
            id: format!("arch-{content_hash}-{}-{}", timestamp_ms, self.seq),
            content_hash,
            residual,
            score,
            reason,
            timestamp_ms,
            generation,
            is_critical_syntactic,
            token_count,
        };
        self.records.push(entry.clone());
        if self.records.len() > self.cap {
            let excess = self.records.len() - self.cap;
            self.records.drain(0..excess);
        }
        entry
    }

    /// Archive a pruned `ContextUnit` (exit from active C).
    pub fn record_unit(
        &mut self,
        unit: &ContextUnit,
        reason: ArchiveReason,
        timestamp_ms: i64,
        generation: Option<u32>,
    ) -> ArchiveRecord {
        self.record(
            unit.content.clone(),
            unit.score,
            reason,
            timestamp_ms,
            generation,
            unit.is_critical_syntactic,
            unit.token_count,
        )
    }

    pub fn size(&self) -> usize {
        self.records.len()
    }

    /// Deliberate retrieval — AND filters, newest first.
    pub fn query(&self, q: &ArchivistQuery) -> Vec<ArchiveRecord> {
        let mut out: Vec<ArchiveRecord> = self
            .records
            .iter()
            .filter(|r| {
                if let Some(ref id) = q.id {
                    if &r.id != id {
                        return false;
                    }
                }
                if let Some(ref h) = q.content_hash {
                    if &r.content_hash != h {
                        return false;
                    }
                }
                if let Some(reason) = q.reason {
                    if r.reason != reason {
                        return false;
                    }
                }
                if let Some(since) = q.since_ms {
                    if r.timestamp_ms < since {
                        return false;
                    }
                }
                if let Some(until) = q.until_ms {
                    if r.timestamp_ms > until {
                        return false;
                    }
                }
                true
            })
            .cloned()
            .collect();
        out.sort_by(|a, b| b.timestamp_ms.cmp(&a.timestamp_ms));
        if let Some(limit) = q.limit {
            out.truncate(limit);
        }
        out
    }

    pub fn snapshot(&self) -> Vec<ArchiveRecord> {
        self.query(&ArchivistQuery::default())
    }

    /// Test helper — clears in-memory log only.
    pub fn clear(&mut self) {
        self.records.clear();
    }
}

impl Default for Archivist {
    fn default() -> Self {
        Self::new()
    }
}

fn simple_hash(s: &str) -> String {
    let mut h: i32 = 0;
    for b in s.bytes() {
        h = h.wrapping_shl(5).wrapping_sub(h).wrapping_add(b as i32);
    }
    format!("h-{:x}", h.unsigned_abs())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn append_only_grows() {
        let mut a = Archivist::new();
        a.record("alpha", 0.2, ArchiveReason::Pruned, 1, None, false, 1);
        a.record("beta", 0.1, ArchiveReason::Budget, 2, Some(1), true, 2);
        assert_eq!(a.size(), 2);
        let snap = a.snapshot();
        assert_eq!(snap[0].residual, "beta"); // newest first
        assert_eq!(snap[1].residual, "alpha");
    }

    #[test]
    fn query_by_reason() {
        let mut a = Archivist::new();
        a.record("a", 0.0, ArchiveReason::Pruned, 1, None, false, 1);
        a.record("b", 0.0, ArchiveReason::Overflow, 2, None, false, 1);
        let q = ArchivistQuery {
            reason: Some(ArchiveReason::Overflow),
            ..Default::default()
        };
        let hits = a.query(&q);
        assert_eq!(hits.len(), 1);
        assert_eq!(hits[0].residual, "b");
    }

    #[test]
    fn cap_bounds_memory() {
        let mut a = Archivist::with_cap(2);
        a.record("1", 0.0, ArchiveReason::Pruned, 1, None, false, 1);
        a.record("2", 0.0, ArchiveReason::Pruned, 2, None, false, 1);
        a.record("3", 0.0, ArchiveReason::Pruned, 3, None, false, 1);
        assert_eq!(a.size(), 2);
        let ids: Vec<_> = a.snapshot().into_iter().map(|r| r.residual).collect();
        assert!(ids.contains(&"2".to_string()));
        assert!(ids.contains(&"3".to_string()));
    }

    #[test]
    fn record_unit_preserves_critical_flag() {
        let mut a = Archivist::new();
        let unit = ContextUnit {
            id: "u1".into(),
            content: "/usr/bin/cargo".into(),
            score: 0.15,
            layer: [0, 1, 2],
            token_count: 1,
            is_critical_syntactic: true,
        };
        let rec = a.record_unit(&unit, ArchiveReason::Pruned, 99, Some(2));
        assert!(rec.is_critical_syntactic);
        assert_eq!(rec.generation, Some(2));
        assert_eq!(rec.score, 0.15);
    }
}
